src/Build/Instance/RunningObjectTable.cs

```diff
@@ -14,34 +16,40 @@ namespace Microsoft.Build.Execution
     /// <remarks>
     /// See https://docs.microsoft.com/en-us/windows/desktop/api/objidl/nn-objidl-irunningobjecttable.
     /// </remarks>
-    internal class RunningObjectTable : IDisposable, IRunningObjectTableWrapper
+    internal class RunningObjectTable : IRunningObjectTableWrapper
     {
-        private readonly IRunningObjectTable rot;
-        private bool isDisposed = false;
+        private readonly Task<IRunningObjectTable> _rotTask;
 
         public RunningObjectTable()
         {
-            Ole32.GetRunningObjectTable(0, out this.rot);
-        }
-
-        public void Dispose()
-        {
-            if (this.isDisposed)
+            if (Thread.CurrentThread.GetApartmentState() == ApartmentState.MTA)
             {
-                return;
+                Ole32.GetRunningObjectTable(0, out var rot);
+                _rotTask = Task.FromResult(rot);
+            }
+            else
+            {
+                // To avoid deadlock, create ROT in a threadpool threads which guarantees to be MTA. And the
+                // object will be MTA
+                _rotTask =
+                Task.Run(() =>
+                    {
+                        Ole32.GetRunningObjectTable(0, out var rot);
+                        return rot;
+                    });
             }
-
-            Marshal.ReleaseComObject(this.rot);
-            this.isDisposed = true;
         }
 
         /// <summary>
         /// Attempts to retrieve an item from the ROT.
         /// </summary>
         public object GetObject(string itemName)
         {
-            IMoniker mk = CreateMoniker(itemName);
-            int hr = this.rot.GetObject(mk, out object obj);
+            Task<IMoniker> mkTask = CreateMoniker(itemName);
+            Task.WaitAll(_rotTask, mkTask);
+            var rot = _rotTask.Result;
```

Comment:
If the `_rotTask` has faulted, then `Task.Result` throws `AggregateException` instead of the original exception (the original is preserved as an inner exception). If you expect exceptions might happen and want to preserve the original, use `_rotTask.GetAwaiter().GetResult()` instead.